### PR TITLE
Use libjpeg-turbo for all Lossless JPEG bit depths

### PR DESCRIPTION
**Pull request by SimonSegerblomRex**
Enabled by the solution to https://github.com/libjpeg-turbo/libjpeg-turbo/issues/768
(Planned to be included in libjpeg-turbo release 3.1.0.)

There are still some Lossless JPEG encoded images that libjpeg-turbo refuses to decode, see the discussions in:
 * https://github.com/libjpeg-turbo/libjpeg-turbo/issues/586
 * https://github.com/libjpeg-turbo/libjpeg-turbo/issues/765

**Note to self while testing with local copy of libjpeg-turbo:**
Put this in the `customize_build` function used by `setup.py`:
```python3
libjpeg_turbo_path = <path to libjpeg-turbo>
EXTENSIONS['jpeg8']['sources'] = []
EXTENSIONS['jpeg8']['include_dirs'] = [libjpeg_turbo_path + "/src"]  # moved to src in the dev branch
EXTENSIONS['jpeg8']['library_dirs'] = [libjpeg_turbo_path]
```
and make sure to set
```sh
export LD_LIBRARY_PATH=<libjpeg_turbo_path>
```
before running any python script importing imagecodecs.

### CODE REVIEW

**Comment by SimonSegerblomRex** [comment]
This is WIP and will stay as a draft pull request until there's an official libjpeg-turbo release that includes the changes necessary.

**Comment by cgohlke** [comment]
Thanks. I am aware of the ongoing work in libjpeg-turbo. Note that the JPEG codec in imagecodecs switches to the [LJPEG codec](https://github.com/cgohlke/imagecodecs/blob/master/imagecodecs/_ljpeg.pyx) for bit-depths not supported by libjpeg-turbo.

**Comment by SimonSegerblomRex** [comment]
> Note that the JPEG codec in imagecodecs switches to the [LJPEG codec](https://github.com/cgohlke/imagecodecs/blob/master/imagecodecs/_ljpeg.pyx) for bit-depths not supported by libjpeg-turbo.

Yes, the existing Lossless JPEG decoders seem to work fine and they will still be needed as backups in `jpeg_decode` for images that libjpeg-turbo refuses to decode due to the issues discussed in https://github.com/libjpeg-turbo/libjpeg-turbo/issues/586 and https://github.com/libjpeg-turbo/libjpeg-turbo/issues/765. `ljpeg_encode` shouldn't be needed any longer though.

File: tests/test_imagecodecs.py [hunk 1]
```diff
@@ -2930,18 +2930,20 @@ def test_ljpeg(fname, result, codec):
...
-            # 'dng2.ljp',  # Bogus Huffman table definition
-            # 'dng3.ljp',  # Bogus Huffman table definition
-            # 'dng4.ljp',  # Bogus Huffman table definition
-            # 'dng5.ljp',  # Bogus Huffman table definition
-            # 'dng6.ljp',  # Bogus Huffman table definition
-            # 'dng7.ljp',  # Bogus Huffman table definition
+            # The below "Bogus Huffman table definition" errors can be fixed
+            # with the suggested patch discussed here:
+            # https://github.com/libjpeg-turbo/libjpeg-turbo/issues/765
+            'dng1.ljp',  # Bogus Huffman table definition
```
Comment (SimonSegerblomRex) [inline tests/test_imagecodecs.py:2940]: @cgohlke, do you know how these "dng*.ljp" image were generated (to understand why the contain the invalid Huffman codes discussed [here](https://github.com/libjpeg-turbo/libjpeg-turbo/issues/765))?

File: imagecodecs/imagecodecs.py [hunk 2]
```diff
@@ -109,7 +109,7 @@
 - `libheif <https://github.com/strukturag/libheif>`_ 1.17.6
   (`libde265 <https://github.com/strukturag/libde265>`_ 1.0.15,
   `x265 <https://bitbucket.org/multicoreware/x265_git/src/master/>`_ 3.6)
-- `libjpeg-turbo <https://github.com/libjpeg-turbo/libjpeg-turbo>`_ 3.0.3
+- `libjpeg-turbo <https://github.com/libjpeg-turbo/libjpeg-turbo>`_ 6ec8e41f50e5a83fe078732cbf0360272165ed45
```
Comment (SimonSegerblomRex) [inline imagecodecs/imagecodecs.py:112]: This is the latest sha1 from the dev branch. No official release tag.

**Comment by SimonSegerblomRex** [comment]
I tested this with a 16bit Lossless JPEG file as input:
```python3
import sys

import numpy as np
from imagecodecs import jpeg8_decode, jpeg8_encode

filename = sys.argv[1]


with open(filename, "rb") as f:
    encoded = f.read()

image = jpeg8_decode(encoded)
if image.ndim > 2:
    image = image[..., 0].copy()  # copy to fix strides

for bit_depth in range(16, 1, -1):
    print(bit_depth)
    image <<= 1
    if bit_depth <= 8:
        # FIXME: Should this really be necessary?
        image = image.astype("u1")
    enc = jpeg8_encode(
        image,
        lossless=True,
        predictor=1,
        bitspersample=bit_depth,
    )
    dec = jpeg8_decode(enc)
    np.testing.assert_array_equal(image, dec)
```
It works, but the case with bit-depth <= 8 in a uint16 array should be handled in a better way.

File: imagecodecs/_jpeg8.pyx [hunk 3]
```diff
@@ -152,8 +152,9 @@ def jpeg8_encode(
 
     if bitspersample is not None:
         if (
-            bitspersample not in {8, 12, 16}
-            or src.itemsize == 1 and bitspersample > 8
+            not (lossless and 2 <= bitspersample <= 16)
+            or (not lossless and bitspersample not in {8, 12})
+            or src.itemsize < (bitspersample + 7) // 8
```
Comment (SimonSegerblomRex) [inline imagecodecs/_jpeg8.pyx:157]: Should this be `!=` instead of `<`? Or how should the case with bitspersample <= 8 stored in uint16 arrays be handled?

Comment (SimonSegerblomRex) [inline imagecodecs/_jpeg8.pyx:157]: Should we use `!=` instead of `<`?

Comment (SimonSegerblomRex) [reply imagecodecs/_jpeg8.pyx]: `!=`

File: imagecodecs/_jpeg8.pyx [hunk 4]
```diff
@@ -141,7 +141,7 @@ def jpeg8_encode(
         (src.dtype == numpy.uint8 or src.dtype == numpy.uint16)
         and src.ndim in {2, 3}
         # src.nbytes <= 2147483647 and  # limit to 2 GB
-        and samples in {1, 3, 4}
+        and samples in {1, 2, 3, 4}
```
Comment (SimonSegerblomRex) [inline imagecodecs/_jpeg8.pyx:144]: Seems to work as expected with 2 components:
```python3
import sys

from imagecodecs import imread, jpeg8_decode, jpeg8_encode
from numpy.testing import assert_array_equal

filename = sys.argv[1]

image = imread(filename)
enc = jpeg8_encode(
    image,
    lossless=True,
    predictor=1,
    bitspersample=16,
)
dec = jpeg8_decode(enc)
assert_array_equal(image, dec)
```
Using this input file:

File: tests/test_imagecodecs.py (hunk 1 above)
Comment (SimonSegerblomRex) [reply tests/test_imagecodecs.py:2939]: ...never mind, I know who's responsible for those images... :sweat_smile: I'll fix them.

**Comment by SimonSegerblomRex** [comment]
(I replaced the broken dng*.ljp files that were created using my broken Lossless JPEG encoder.)

Everything seems to work as expected now, but I guess we should wait for an official libjpeg-turbo tag.

**Comment by SimonSegerblomRex** [comment]
I found [this source](https://sourceforge.net/projects/gdcm/files/gdcmData/gdcmData/gdcmData.tar.gz/download) containing a lot of Lossless JPEG files (embedded in DICOM files). A quick test shows that libjpeg-turbo and lj92 produce slightly different results for some of them, e.g., **gdcm-JPEG-LossLessThoravision.dcm**.

**Comment by SimonSegerblomRex** [comment]
> @SimonSegerblomRex What do you get if you use thorfdbg/libjpeg ?

With thorfdbg/libjpeg I get:
```
reading a JPEG file failed - error -1038 - invalid stream, found invalid huffman code in entropy coded segment
```
and that's probably the right thing. The images decoded by lj92 and libjpeg-turbo are completely broken, so they probably should fail as well.

**Comment by SimonSegerblomRex** [comment]
I found that lj92 fails to decode **MARCONI_MxTWin-12-MONO2-JpegLossless-ZeroLengthSQ.dcm** (just 0s out) while libjpeg-turbo decodes it without issues :+1:

**Comment by SimonSegerblomRex** [comment]
> What kind of command did you use ?

I just extracted the JPEG file looking for the markers:  MARCONI-extracted

```console
> jpeg MARCONI-extracted.jpg out.pgm
jpeg Copyright (C) 2012-2018 Thomas Richter, University of Stuttgart
and Accusoft

For license conditions, see README.license for details.

reading a JPEG file failed - error -1038 - invalid stream, found invalid huffman code in entropy coded segment

0 bytes memory not yet released.

28891 bytes maximal required.

54 allocations performed.
```
This is using the latest version (1.69). Did you get any useful output?

**Comment by SimonSegerblomRex** [comment]
This is ready for code review (but there's till no new libjpeg-turbo release or tag).

**Comment by cgohlke** [comment]
Thank you. I will review this when libjpeg-turbo 3.1 is released.

**Comment by cgohlke** [comment]
I have tested this with libjpeg-turbo 3.1 beta and it works as expected. The changes will be in the next release of imagecodecs along with some tweaks to make tests pass with libjpeg-turbo 3.0. Thank you.

**Comment by SimonSegerblomRex** [comment]
Thank you @cgohlke! Do you think you'll have time to publish the new release soon? If not it would be great if you could push your changes to a dev branch, I would like to verify that my main use-case (with two components) still works.

**Comment by cgohlke** [comment]
The plan is to do a release before Python 3.13, this or next weekend. Hopefully there are no major issues on macOS.

I am attaching the current [_jpeg8.pyx](https://github.com/user-attachments/files/17077959/_jpeg8.zip), which should be enough for you to test, no?

**Comment by SimonSegerblomRex** [comment]
> The plan is to do a release before Python 3.13, this or next weekend. Hopefully there are no major issues on macOS.

Thank you, sounds good!


> I am attaching the current [_jpeg8.pyx](https://github.com/user-attachments/files/17077959/_jpeg8.zip), which should be enough for you to test, no?

Yes, I confirmed that it works 👍

**Comment by cgohlke** [comment]
Fixed in imagecodecs 2024.9.22.

**Comment by SimonSegerblomRex** [comment]
> Fixed in imagecodecs 2024.9.22.

Thank you for the new tag @cgohlke! It works well when I build imagecodecs from source against libjpeg-turbo 3.0.90beta. The wheel release was built against libjpeg-turbo 3.0.4, right? Things like encoding images with 14 bits bitspersample as 2 components still doesn't work when installing the wheel. Will you consider making a release built against libjpeg-turbo 3.0.90beta, or will you wait for the 3.1 release?

**Comment by cgohlke** [comment]
The released wheels are built against libjpeg-turbo 3.0.4. I'll wait for version 3.1. No, the system JPEG library is not used. The dynamic JPEG library used is in the `imagecodecs/libs` directory, with a name like `libjpeg-3a8ca8f3.so.8`.

**Pull request closed without merging**